### DISCUSSION TITLE
Does it work with V2 gateway payloads?

### DISCUSSION THREAD

**Discussion by suceava** [dbartholomae/lambda-middleware#66] [Q&A]
Currently `compose()` passes in a `APIGatewayProxyEvent` but I'm working with `APIGatewayProxyEventV2`.  Is there an option to make it use the V2 payload?
For context here are the 2 versions: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html

I've already built my API around V2 so I'm hoping I don't have to change it.  Thanks for any help.

**Comment by dbartholomae** [comment]
Compose itself does not rely on any specific type. Which middlewares are you using? Can you maybe share the relevant code?

**Reply by suceava** [reply]
I'm looking at using `jwt-auth` and `http-error-handler` to start with.
I just noticed `compose` uses generic types, so I was able to work out the types needed for it, but digging into the code for `jwt-auth` and `http-error-handler` it seems they are specifically using V1 types.
Here's my relevant code that I'm trying out (I've aliased the types as I was working them out in my head):
```Typescript
const getById = async (event: APIGatewayProxyEventV2) => {
}

type T1 = APIGatewayProxyEventV2 & AuthorizedEvent<unknown>;
type T2 = APIGatewayProxyResultV2;
type T3 = PromiseHandler<T1, T2>;
type T4 = (event: APIGatewayProxyEventV2, context: Context) => Promise<any>;
export const handler: APIGatewayProxyHandlerV2 = compose<[T3], T4>(
  jwtAuth({
    algorithm: EncryptionAlgorithms.HS256,
    credentialsRequired: true,
    secretOrPublicKey: process.env.JWT_SECRET || "",
  })
)(getById);
```

**Comment by dbartholomae** [comment]
Thanks! The ideal solution would be to have the middlewares define just what they need from the input event and use generics to push it through to the output, but, unfortunately, TypeScript can't handle generics in a compose function properly. At least it couldn't when I last tried.
One alternative that comes to my mind would be to use function overloading to allow either `APIGatewayProxyEvent` with an `APIGatewayProxyResult`, or `APIGatewayProxyEvent2` with an `APIGatewayProxyResult2`.
I'm open to pull requests around this idea.
